drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1181 @@
+			for (iter = output_list; iter; iter = iter->next) {
+				if (iter->crtc_id == crtc) {
+					crtc = 0;
+					break;
+				}
+			}
+
+			if (crtc > 0) {
+				drmModeFreeEncoder(enc);
+				out->crtc_id = crtc;
+				return 0;
+			}
+		}
+
+		drmModeFreeEncoder(enc);
+	}
+
+	/* If the connector is not currently bound to an encoder or if the
+	 * encoder+crtc is already used by another connector (actually unlikely
+	 * but lets be safe), iterate all other available encoders to find a
+	 * matching CRTC.
+	 */
+	for (i = 0; i < conn->count_encoders; ++i) {
+		enc = drmModeGetEncoder(fd, conn->encoders[i]);
+		if (!enc) {
+			fprintf(stderr, "cannot retrieve encoder %u:%u (%d): %m\n",
+				i, conn->encoders[i], errno);
+			continue;
+		}
+
+		/* iterate all global CRTCs */
+		for (j = 0; j < res->count_crtcs; ++j) {
+			/* check whether this CRTC works with the encoder */
+			if (!(enc->possible_crtcs & (1 << j)))
+				continue;
+
+			/* check that no other output already uses this CRTC */
+			crtc = res->crtcs[j];
+			for (iter = output_list; iter; iter = iter->next) {
+				if (iter->crtc_id == crtc) {
+					crtc = 0;
+					break;
+				}
+			}
+
+			/* We have found a CRTC, so save it and return. Note that we have to
+			 * save its index as well. The CRTC index (not its ID) will be used
+			 * when searching for a suitable plane. TODO: the index is not
+			 * really persistent?
+			 */
+			if (crtc > 0) {
+				fprintf(stdout, "crtc %u found for encoder %u, will need full modeset\n",
+					crtc, conn->encoders[i]);;
+				drmModeFreeEncoder(enc);
+				out->crtc_id = crtc;
+				out->crtc_index = j;
+				return 0;
+			}
+		}
+
+		drmModeFreeEncoder(enc);
+	}
+
+	fprintf(stderr, "cannot find suitable crtc for connector %u\n",
+		conn->connector_id);
+	return -ENOENT;
+}
+
+/*
+ * modeset_find_plane() is a new function. Given a certain combination
+ * of connector+CRTC, it looks for a primary plane for it.
+ */
+
+static int modeset_find_plane(int fd, struct modeset_output *out)
+{
+	drmModePlaneResPtr plane_res;
+	bool found_primary = false;
+	int i, ret = -EINVAL;
+
+	plane_res = drmModeGetPlaneResources(fd);
+	if (!plane_res) {
+		fprintf(stderr, "drmModeGetPlaneResources failed: %s\n",
+				strerror(errno));
+		return -ENOENT;
+	}
+
+	/* iterates through all planes of a certain device */
+	for (i = 0; (i < plane_res->count_planes) && !found_primary; i++) {
+		int plane_id = plane_res->planes[i];
+
+		drmModePlanePtr plane = drmModeGetPlane(fd, plane_id);
+		if (!plane) {
+			fprintf(stderr, "drmModeGetPlane(%u) failed: %s\n", plane_id,
+					strerror(errno));
+			continue;
+		}
+
+		/* check if the plane can be used by our CRTC */
+		if (plane->possible_crtcs & (1 << out->crtc_index)) {
+			drmModeObjectPropertiesPtr props =
+				drmModeObjectGetProperties(fd, plane_id, DRM_MODE_OBJECT_PLANE);
+
+			/* get the "type" property and check if this is a primary plane */
+			if (get_property_value(fd, props, "type") == DRM_PLANE_TYPE_PRIMARY) {
```

Comment:
Hmmm, I didn't get it.

Reply:
`DRM_PLANE_TYPE_PRIMARY` is special: it is defined in the kernel's UAPI headers that userspace use. That is not generally true for all properties, especially for properties of type enum.

Compare to "HDCP Content Type" property (https://dri.freedesktop.org/docs/drm/gpu/drm-kms.html#standard-connector-properties). You will not find `DRM_MODE_HDCP_CONTENT_TYPE0` defined in any userspace visible header. So how do you set content type to type0? You have to ask the kernel what value the string `"HDCP Type0"` corresponds to. Then set the property to that value. You find the string<->value mapping via `drmModeGetProperty()`.

Why are `"type"` property and several other properties so special that their enum values can be found directly in UAPI headers? Legacy. The kernel can never remove the definitions because it might break userspace, but userspace is still supposed use the string names primarily and convert to numerical values at runtime. This is done for UAPI stability reasons: string names are easier to both make unique and keep consistent between drivers and kernel versions.

Reply:
Oh now I get it. Thanks a lot for detailed explanation!